### PR TITLE
Enable duplicated symbols location output

### LINKED ISSUE #150: The cause duplicate symbols is not completly displayed.
First I use `xcodebuild` to build my project, duplicated symbols are displayed with full path:

```
ld: warning: duplicate symbol '_ZSTD_copyCCtx' in:
    /Volumes/test_result/Build/Products/DailyBuild-iphonesimulator/XCFrameworkIntermediates/Btts/libbtts.a(zstd_compress.o)
    /Volumes/est_result/Build/Products/DailyBuildiphonesimulator/XCFrameworkIntermediates/Diagnose/DefaultLogger/mars.framework/mars(zstd_compress.c.o)
```

I use `xcbeautify` to filter my `xcodebuild` logs like below:

```
set -o pipefail && xcodebuild build-for-testing -workspace xxx -quiet | xcbeautify -qq


❌ duplicate symbol '_ZSTD_copyCCtx' in
❌ duplicate symbol '_ZSTD_getCParams' in
```

Only symbol name is printed. That's not easy to findout where are the duplicated objects.

I think it is a bug.

### PR DESCRIPTION
**Pull request by blastmann**
Try to fix #150.

1. Support printing `LinkerDuplicateSymbolsLocation` and `LinkerUndefinedSymbolLocation` in Terminal and GithubActionsLog.
2. Fix `ldWarning` matching.  The original regex of `warning` will match `ldWarning` also, then message output is hidden sometimes.
3. Level up `undefinedSymbolLocation` to error.

### CODE REVIEW

File: Sources/XcbeautifyLib/String+CapturedGroups.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/String+CapturedGroups.swift]: Looks like there's a mismatch between the parameter and the variable. Can you update to reflect the actual expected value?

File: Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift]: I think this should use `Format.indent` instead of leading spaces.

File: Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift]: Looks like you're supplying the wrong arguments here. You'll want something something like...

```swift
let fileComponents = group.filePath.asFileComponents()
let message = group.message // or whatever other property holds the warning

return outputGitHubActionsLog(
    annotationType: .warning,
    fileComponents: fileComponents,
    message: message
)
```

File: Sources/XcbeautifyLib/Pattern.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Pattern.swift:317]: Just curious, why does this exclude `ld`?

**Review by cpisciotta — CHANGES_REQUESTED** [review]
Hey @blastmann! Thanks for improving xcbeautify! Looks like we need to clean up a few things before this is ready to merge.

As part of this PR, can you please add some unit tests against your changes?

Let me know if you have any questions!

File: Sources/XcbeautifyLib/Pattern.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Pattern.swift:317]: Here is an example.

The original regex will also match build log like this (Use Xcode 14.x or enable `-ld64` under Xcode 15) :

```
ld: warning: Could not find or use auto-linked framework 'XYFramework'
Undefined symbols for architecture x86_64:
  "__swift_FORCE_LOAD_$_swiftCompatibility56", referenced from:
      __swift_FORCE_LOAD_$_swiftCompatibility56_$_ABCKit in ABCKit(y.o)
```

Actually, that's not a compile warning and then `formatCompileWarning` will read 2 more addtional lines. So the actual errors will be suppressed.

So I exclude `ld` and print it correctly with `formatLdWarning`.

File: Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift]: The warning message here is `filePath` only.

Can we leave the message empty here?

File: Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift]: Got it.

File: Sources/XcbeautifyLib/String+CapturedGroups.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/String+CapturedGroups.swift]: OK.

File: Sources/XcbeautifyLib/Pattern.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Pattern.swift:317]: Looks like there is a matching order issue with `ld` related regex expressions.

Once it matches `ldError`, next time `ldWarning` can't be matched as usual. I think I should create a new `ld` related issue and fix it later.

**Comment by blastmann** [comment]
> Hey @blastmann! Thanks for improving xcbeautify! Looks like we need to clean up a few things before this is ready to merge.
> 
> As part of this PR, can you please add some unit tests against your changes?
> 
> Let me know if you have any questions!

Unit test is added. Check please.

File: Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift]: This is defined as an error in `Parser.swift`. Do you know which we want?

https://github.com/tuist/xcbeautify/blob/4bbfdc863e175f347674ef0a60ab2048078edea6/Sources/XcbeautifyLib/Parser.swift#L79

File: Sources/XcbeautifyLib/Parser.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Parser.swift:90]: Hmmm... It's before I started working on this project, but I'm wondering, what's the difference between this and `.linkerUndefinedSymbolLocation`? If both are needed, we should probably distinguish the two with documentation comments.

https://github.com/tuist/xcbeautify/blob/4bbfdc863e175f347674ef0a60ab2048078edea6/Sources/XcbeautifyLib/Parser.swift#L81

File: Sources/XcbeautifyLib/Pattern.swift
Comment (cpisciotta) [reply Sources/XcbeautifyLib/Pattern.swift:317]: Got it. Thanks for the clarification!

File: Sources/XcbeautifyLib/Pattern.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Pattern.swift:322]: Do you know if we can avoid the catchall (`.*`) here and on the modifications below? I recognize this is used by many of the other regexes, and there may not be a reliable alternative. That said, this has introduced some performance issues in the past, and it'd be helpful to remove them while we're here, if possible.

File: Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift]: I think we want to prefix `Symbol.error` and `Symbol.asciiError` to these messages, right?

File: Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift
Comment (cpisciotta) [reply Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift]: If that's the case, then I think we'd want to parse the warning as a message rather than a `filePath`.

File: Sources/XcbeautifyLib/Parser.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Parser.swift:90]: I'm confused too. Looks like both of these error mesages appear at the stage of linking, but they display in different ways.

File: Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift]: I think this should be defined as an error also. I will fix it later.

File: Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift]: I think error messages with symbol prefixes is a little bit distractive. But prefix it with a symbol can be more integrity.

I'll fix it.

File: Sources/XcbeautifyLib/Pattern.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/Pattern.swift:322]: I think it should become another issue.

`ld` related messages are become unstable in recent Xcode. We can enhance the performance later.

File: Tests/XcbeautifyLibTests/XCTestManifests.swift
Comment (cpisciotta) [inline Tests/XcbeautifyLibTests/XCTestManifests.swift:1]: IIRC, some of the tests you introduced in this file have been excluded, since they've had known problems. You'll need to revert some of these changes to unblock your PR, but I agree they should eventually be part of this file.

File: Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift
Comment (cpisciotta) [inline Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift:289]: Since this is a warning, the prefix should be `[!]`.

File: Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift
Comment (cpisciotta) [inline Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift:310]: nit: Can you break these into separate tests? They're testing 3 independent inputs and outputs.

File: Sources/XcbeautifyLib/String+CapturedGroups.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/String+CapturedGroups.swift]: The variable name and argument label no longer match. We should fix them here (or the underlying logic) to avoid the discrepency.

File: Sources/XcbeautifyLib/CaptureGroups.swift
Comment (cpisciotta) [inline Sources/XcbeautifyLib/CaptureGroups.swift]: Why did you change this to `wholeError`? That seems unexpected based on the type. We already have a `CompileErrorCaptureGroup` that we should use for compiler errors.

**Review by cpisciotta — COMMENTED** [review]
Just a couple of additional housekeeping items, and it looks like we need to address some Ubuntu/Linux test failures.

File: Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift
Comment (blastmann) [reply Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift:289]: This message error level has been upgrade to `ERROR`.

Here is another problem I didn't solve: I can't find an elegant way to detect where this build log comes from. There's no information about ld version in build log.

But I think `duplicate symbol` messages are important, so I level up this message level.

File: Sources/XcbeautifyLib/CaptureGroups.swift
Comment (blastmann) [reply Sources/XcbeautifyLib/CaptureGroups.swift]: It is renamed by Xcode.

That should be reverted.

File: Tests/XcbeautifyLibTests/XCTestManifests.swift
Comment (blastmann) [reply Tests/XcbeautifyLibTests/XCTestManifests.swift:1]: This manifest file has a compile issue last time. I have to solve it by running `swift test --generate-linuxmain`.

Do you recommend I should delete these lines of update?

File: Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift
Comment (blastmann) [reply Tests/XcbeautifyLibTests/RendererTests/TerminalRendererTests.swift:310]: This is not three separate test cases. The original logic adjusted the priority of `ldError` after processing `#"^(ld:.*)"#`, causing `ldWarning` to never be hit. Therefore, I separately verified the hit logic for ldError in the test.

**Comment by pepicrft** [comment]
What's the state of this? We'll have to rebase the most recent changes in `main`.

**Comment by blastmann** [comment]
> What's the state of this? We'll have to rebase the most recent changes in `main`.

Actually, I've done what I can do in this PR. If anyone has any comments about this change list, please let me know.

**Pull request closed without merging**
